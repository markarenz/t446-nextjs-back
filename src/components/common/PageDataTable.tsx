import Link from 'next/link';
import { useRouter } from 'next/router';
import { Page, Alert, Gallery, FieldDef } from '../../types/types';
import IconButton from '../../components/common/IconButton';
import { IconEdit, IconView, IconDelete } from '../../components/img/icons';

// handleNavigateToEdit
type Props = {
  items: Page[] | Alert[] | Gallery[]; // Add other types
  tableFields: FieldDef[];
  slug: string;
  tableActions: string[];
  handleDelete: Function;
};
const PageDataTable: React.FC<Props> = ({
  items,
  slug,
  tableFields,
  tableActions,
  handleDelete
}) => {
  const router = useRouter();
  const handleNavigateToEdit = (id: string): void => {
    router.replace(`/${slug}/edit/${id}`);
  };

  return (
    <table className="mb-2">
      <thead>
        <tr>
          {tableFields.map((f) => (
            <th key={f.title}>{f.title}</th>
          ))}
          <th>&nbsp;</th>
        </tr>
      </thead>
      <tbody>
        {items?.map((i) => (
          <tr key={`${i.id}`}>
            <td>
              <label>Title:</label>
              <Link href={`pages/edit/${i.id}`}>{i.title}</Link>
            </td>
            <td>
              <label>Status:</label>
              {i.status.toUpperCase()}
            </td>
            <td className="right">
              {tableActions.includes('edit') && (
                <span className="mr-1">
                  <IconButton
                    onClick={() => handleNavigateToEdit(`${i.id}`)}
                    title="edit"
                    color="secondary"
                  >
                    <IconEdit />
                  </IconButton>
                </span>
              )}
              {tableActions.includes('view') && (
                <span className="mr-1">
                  <a
                    // @ts-ignore
                    href={`https://www.indytroop446.org/${i?.slug}`}
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    <IconButton onClick={() => {}} title="view">
                      <IconView />
                    </IconButton>
                  </a>
                </span>
              )}
              {tableActions.includes('delete') && (
                <span>
                  <IconButton
                    onClick={() => handleDelete(i)}
                    title="delete"
                    color="primary"
                  >
                    <IconDelete />
                  </IconButton>
                </span>
              )}
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default PageDataTable;
