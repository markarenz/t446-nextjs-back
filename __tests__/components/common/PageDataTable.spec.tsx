import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import PageDataTable from '../../../src/components/common/PageDataTable';
import { Alert } from '@prisma/client';

const replaceMock = jest.fn();
jest.mock('next/router', () => ({
  useRouter: jest.fn(() => ({
    replace: replaceMock
  }))
}));

const tableActions = ['edit', 'view', 'delete'];
const tableFields = [
  {
    title: 'Title',
    slug: 'title'
  },
  {
    title: 'Modified',
    slug: 'dateModified'
  },
  {
    title: 'Status',
    slug: 'status'
  }
];
const defaultDate = new Date('2022-12-01');
const items: Alert[] = [
  {
    id: 'abcd123',
    title: 'test 1',
    status: 'active',
    content: 'This is test **content**.',
    alwaysOn: false,
    dateStart: defaultDate,
    dateEnd: defaultDate,
    dateCreated: defaultDate,
    dateModified: defaultDate
  },
  {
    id: 'abcd124',
    title: 'test 2',
    status: 'active',
    content: 'This is test **content**.',
    alwaysOn: false,
    dateStart: defaultDate,
    dateEnd: defaultDate,
    dateCreated: defaultDate,
    dateModified: defaultDate
  }
];

const defaultProps = {
  tableFields,
  slug: 'alerts',
  tableActions,
  items,
  handleDelete: jest.fn(),
  viewPrefix: 'https://'
};
describe('PageDataTable', () => {
  it('renders component', async () => {
    render(<PageDataTable {...defaultProps} />);
    const element = await screen.findByTestId('page-data-table');
    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });
  it('redirects to edit page when btn clicked', async () => {
    render(<PageDataTable {...defaultProps} />);
    const element = await screen.findByTestId('page-table-item-abcd124');
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(replaceMock).toHaveBeenCalled();
  });

  it('calls delete function when delete btn clicked', async () => {
    render(<PageDataTable {...defaultProps} />);
    const btnView = await screen.findByTestId('page-table-view-item-abcd124');
    fireEvent(
      btnView,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );

    const element = await screen.findByTestId('page-table-delete-item-abcd124');
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(defaultProps.handleDelete).toHaveBeenCalled();
  });
});
