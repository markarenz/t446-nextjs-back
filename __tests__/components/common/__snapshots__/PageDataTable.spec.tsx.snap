// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageDataTable renders component 1`] = `
<table
  class="mb-2"
  data-testid="page-data-table"
>
  <thead>
    <tr>
      <th>
        Title
      </th>
      <th>
        Modified
      </th>
      <th>
        Status
      </th>
      <th>
        Â 
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <label>
          Title:
        </label>
        <a
          href="/edit/abcd123"
        >
          test 1
        </a>
      </td>
      <td>
        <label>
          Modified:
        </label>
        <span>
          11/30/2022, 7:00:00 PM
        </span>
      </td>
      <td>
        <label>
          Status:
        </label>
        <span>
          ACTIVE
        </span>
      </td>
      <td
        class="right"
      >
        <span
          class="mr-1"
        >
          <button
            class="iconButton secondary"
            data-testid="page-table-item-abcd123"
            title="edit"
            type="button"
          >
            <div
              class="inner"
            >
              <svg
                version="1.1"
                viewBox="0 0 490.273 490.273"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M313.548 152.387l-230.8 230.9c-6.7 6.7-6.7 17.6 0 24.3 3.3 3.3 7.7 5 12.1 5s8.8-1.7 12.1-5l230.8-230.8c6.7-6.7 6.7-17.6 0-24.3-6.6-6.8-17.5-6.8-24.2-.1zM431.148 191.887c4.4 0 8.8-1.7 12.1-5l25.2-25.2c29.1-29.1 29.1-76.4 0-105.4l-34.4-34.4c-14.1-14.1-32.8-21.8-52.7-21.8-19.9 0-38.6 7.8-52.7 21.8l-25.2 25.2c-6.7 6.7-6.7 17.6 0 24.3l115.6 115.6c3.3 3.2 7.7 4.9 12.1 4.9zm-78.2-145.9c7.6-7.6 17.7-11.8 28.5-11.8 10.7 0 20.9 4.2 28.5 11.8l34.4 34.4c15.7 15.7 15.7 41.2 0 56.9l-13.2 13.2-91.4-91.4 13.2-13.1z"
                />
                <path
                  d="M162.848 467.187l243.5-243.5c6.7-6.7 6.7-17.6 0-24.3s-17.6-6.7-24.3 0l-239.3 239.5-105.6 14.2 14.2-105.6 228.6-228.6c6.7-6.7 6.7-17.6 0-24.3-6.7-6.7-17.6-6.7-24.3 0l-232.6 232.8c-2.7 2.7-4.4 6.1-4.9 9.8l-18 133.6c-.7 5.3 1.1 10.6 4.9 14.4 3.2 3.2 7.6 5 12.1 5 .8 0 1.5-.1 2.3-.2l133.6-18c3.7-.4 7.2-2.1 9.8-4.8z"
                />
              </svg>
            </div>
          </button>
        </span>
        <span
          class="mr-1"
        >
          <a
            href="https://www.indytroop446.org/https://undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            <button
              class="iconButton default"
              data-testid="page-table-view-item-abcd123"
              title="view"
              type="button"
            >
              <div
                class="inner"
              >
                <svg
                  version="1.1"
                  viewBox="0 0 80.794 80.794"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M79.351 38.549c-.706-.903-17.529-22.119-38.953-22.119-21.426 0-38.249 21.216-38.955 22.119L0 40.396l1.443 1.847c.706.903 17.529 22.12 38.955 22.12 21.424 0 38.247-21.217 38.953-22.12l1.443-1.847-1.443-1.847zM40.398 58.364c-15.068 0-28.22-13.046-32.643-17.967 4.425-4.922 17.576-17.966 32.643-17.966 15.066 0 28.218 13.045 32.642 17.966-4.426 4.922-17.577 17.967-32.642 17.967z"
                  />
                  <path
                    d="M40.397 23.983c-9.052 0-16.416 7.363-16.416 16.414 0 9.053 7.364 16.417 16.416 16.417s16.416-7.364 16.416-16.417c0-9.051-7.364-16.414-16.416-16.414zm0 26.83c-5.744 0-10.416-4.673-10.416-10.417 0-5.742 4.672-10.414 10.416-10.414 5.743 0 10.416 4.672 10.416 10.414 0 5.744-4.673 10.417-10.416 10.417z"
                  />
                </svg>
              </div>
            </button>
          </a>
        </span>
        <span>
          <button
            class="iconButton primary"
            data-testid="page-table-delete-item-abcd123"
            title="delete"
            type="button"
          >
            <div
              class="inner"
            >
              <svg
                version="1.1"
                viewBox="0 0 297.001 297.001"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M220.879 138.078l2.194-36.772h6.967c5.458 0 9.884-4.425 9.884-9.884V67.619c0-19.301-15.703-35.004-35.004-35.004h-30.227V9.884c0-5.458-4.425-9.884-9.884-9.884H99.578c-5.458 0-9.884 4.425-9.884 9.884v22.732H59.466c-19.301 0-35.004 15.703-35.004 35.004v23.803c0 5.458 4.425 9.884 9.884 9.884h6.967l11.124 186.399a9.884 9.884 0 009.866 9.295h139.78a9.884 9.884 0 009.866-9.295l1.938-32.467c32.355-.085 58.652-26.429 58.652-58.804 0-30.006-22.594-54.818-51.66-58.357zM109.461 19.767h45.464v12.849h-45.464V19.767zM44.23 81.539v-13.92c0-8.401 6.835-15.236 15.236-15.236h145.453c8.402 0 15.237 6.835 15.237 15.236v13.92H44.23zM203.241 101.8l-2.221 37.228a58.347 58.347 0 00-10.02 3.175V101.8h12.241zM73.386 277.233h-1.772L61.145 101.8h12.241v175.433zm31.451 0H93.153V101.306h11.684v175.927zm34.945 0h-15.179V101.8h15.179v175.433zm19.767-175.927h11.684v54.547c-4.892 5.121-8.882 11.11-11.684 17.725v-72.272zm11.685 175.928H159.55V219.293c2.801 6.615 6.792 12.604 11.684 17.725v40.216zm21.539-.001H191v-26.566c1.077.453 2.174.865 3.283 1.255l-1.51 25.311zm20.959-41.758c-21.527 0-39.04-17.513-39.04-39.04 0-21.527 17.513-39.04 39.04-39.04 21.527 0 39.04 17.513 39.04 39.04 0 21.527-17.513 39.04-39.04 39.04z"
                />
                <path
                  d="M237.029 173.138a9.886 9.886 0 00-13.978 0l-9.319 9.319-9.319-9.319a9.886 9.886 0 00-13.978 0c-3.86 3.86-3.86 10.118 0 13.978l9.319 9.319-9.319 9.319c-3.86 3.86-3.86 10.118 0 13.978a9.856 9.856 0 006.989 2.895c2.53 0 5.059-.965 6.989-2.895l9.319-9.319 9.319 9.319a9.856 9.856 0 006.989 2.895c2.53 0 5.059-.965 6.989-2.895 3.86-3.86 3.86-10.118 0-13.978l-9.319-9.319 9.319-9.319c3.86-3.86 3.86-10.118 0-13.978z"
                />
              </svg>
            </div>
          </button>
        </span>
      </td>
    </tr>
    <tr>
      <td>
        <label>
          Title:
        </label>
        <a
          href="/edit/abcd124"
        >
          test 2
        </a>
      </td>
      <td>
        <label>
          Modified:
        </label>
        <span>
          11/30/2022, 7:00:00 PM
        </span>
      </td>
      <td>
        <label>
          Status:
        </label>
        <span>
          ACTIVE
        </span>
      </td>
      <td
        class="right"
      >
        <span
          class="mr-1"
        >
          <button
            class="iconButton secondary"
            data-testid="page-table-item-abcd124"
            title="edit"
            type="button"
          >
            <div
              class="inner"
            >
              <svg
                version="1.1"
                viewBox="0 0 490.273 490.273"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M313.548 152.387l-230.8 230.9c-6.7 6.7-6.7 17.6 0 24.3 3.3 3.3 7.7 5 12.1 5s8.8-1.7 12.1-5l230.8-230.8c6.7-6.7 6.7-17.6 0-24.3-6.6-6.8-17.5-6.8-24.2-.1zM431.148 191.887c4.4 0 8.8-1.7 12.1-5l25.2-25.2c29.1-29.1 29.1-76.4 0-105.4l-34.4-34.4c-14.1-14.1-32.8-21.8-52.7-21.8-19.9 0-38.6 7.8-52.7 21.8l-25.2 25.2c-6.7 6.7-6.7 17.6 0 24.3l115.6 115.6c3.3 3.2 7.7 4.9 12.1 4.9zm-78.2-145.9c7.6-7.6 17.7-11.8 28.5-11.8 10.7 0 20.9 4.2 28.5 11.8l34.4 34.4c15.7 15.7 15.7 41.2 0 56.9l-13.2 13.2-91.4-91.4 13.2-13.1z"
                />
                <path
                  d="M162.848 467.187l243.5-243.5c6.7-6.7 6.7-17.6 0-24.3s-17.6-6.7-24.3 0l-239.3 239.5-105.6 14.2 14.2-105.6 228.6-228.6c6.7-6.7 6.7-17.6 0-24.3-6.7-6.7-17.6-6.7-24.3 0l-232.6 232.8c-2.7 2.7-4.4 6.1-4.9 9.8l-18 133.6c-.7 5.3 1.1 10.6 4.9 14.4 3.2 3.2 7.6 5 12.1 5 .8 0 1.5-.1 2.3-.2l133.6-18c3.7-.4 7.2-2.1 9.8-4.8z"
                />
              </svg>
            </div>
          </button>
        </span>
        <span
          class="mr-1"
        >
          <a
            href="https://www.indytroop446.org/https://undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            <button
              class="iconButton default"
              data-testid="page-table-view-item-abcd124"
              title="view"
              type="button"
            >
              <div
                class="inner"
              >
                <svg
                  version="1.1"
                  viewBox="0 0 80.794 80.794"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M79.351 38.549c-.706-.903-17.529-22.119-38.953-22.119-21.426 0-38.249 21.216-38.955 22.119L0 40.396l1.443 1.847c.706.903 17.529 22.12 38.955 22.12 21.424 0 38.247-21.217 38.953-22.12l1.443-1.847-1.443-1.847zM40.398 58.364c-15.068 0-28.22-13.046-32.643-17.967 4.425-4.922 17.576-17.966 32.643-17.966 15.066 0 28.218 13.045 32.642 17.966-4.426 4.922-17.577 17.967-32.642 17.967z"
                  />
                  <path
                    d="M40.397 23.983c-9.052 0-16.416 7.363-16.416 16.414 0 9.053 7.364 16.417 16.416 16.417s16.416-7.364 16.416-16.417c0-9.051-7.364-16.414-16.416-16.414zm0 26.83c-5.744 0-10.416-4.673-10.416-10.417 0-5.742 4.672-10.414 10.416-10.414 5.743 0 10.416 4.672 10.416 10.414 0 5.744-4.673 10.417-10.416 10.417z"
                  />
                </svg>
              </div>
            </button>
          </a>
        </span>
        <span>
          <button
            class="iconButton primary"
            data-testid="page-table-delete-item-abcd124"
            title="delete"
            type="button"
          >
            <div
              class="inner"
            >
              <svg
                version="1.1"
                viewBox="0 0 297.001 297.001"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M220.879 138.078l2.194-36.772h6.967c5.458 0 9.884-4.425 9.884-9.884V67.619c0-19.301-15.703-35.004-35.004-35.004h-30.227V9.884c0-5.458-4.425-9.884-9.884-9.884H99.578c-5.458 0-9.884 4.425-9.884 9.884v22.732H59.466c-19.301 0-35.004 15.703-35.004 35.004v23.803c0 5.458 4.425 9.884 9.884 9.884h6.967l11.124 186.399a9.884 9.884 0 009.866 9.295h139.78a9.884 9.884 0 009.866-9.295l1.938-32.467c32.355-.085 58.652-26.429 58.652-58.804 0-30.006-22.594-54.818-51.66-58.357zM109.461 19.767h45.464v12.849h-45.464V19.767zM44.23 81.539v-13.92c0-8.401 6.835-15.236 15.236-15.236h145.453c8.402 0 15.237 6.835 15.237 15.236v13.92H44.23zM203.241 101.8l-2.221 37.228a58.347 58.347 0 00-10.02 3.175V101.8h12.241zM73.386 277.233h-1.772L61.145 101.8h12.241v175.433zm31.451 0H93.153V101.306h11.684v175.927zm34.945 0h-15.179V101.8h15.179v175.433zm19.767-175.927h11.684v54.547c-4.892 5.121-8.882 11.11-11.684 17.725v-72.272zm11.685 175.928H159.55V219.293c2.801 6.615 6.792 12.604 11.684 17.725v40.216zm21.539-.001H191v-26.566c1.077.453 2.174.865 3.283 1.255l-1.51 25.311zm20.959-41.758c-21.527 0-39.04-17.513-39.04-39.04 0-21.527 17.513-39.04 39.04-39.04 21.527 0 39.04 17.513 39.04 39.04 0 21.527-17.513 39.04-39.04 39.04z"
                />
                <path
                  d="M237.029 173.138a9.886 9.886 0 00-13.978 0l-9.319 9.319-9.319-9.319a9.886 9.886 0 00-13.978 0c-3.86 3.86-3.86 10.118 0 13.978l9.319 9.319-9.319 9.319c-3.86 3.86-3.86 10.118 0 13.978a9.856 9.856 0 006.989 2.895c2.53 0 5.059-.965 6.989-2.895l9.319-9.319 9.319 9.319a9.856 9.856 0 006.989 2.895c2.53 0 5.059-.965 6.989-2.895 3.86-3.86 3.86-10.118 0-13.978l-9.319-9.319 9.319-9.319c3.86-3.86 3.86-10.118 0-13.978z"
                />
              </svg>
            </div>
          </button>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
